Add regional roles, rework appearance of r!roles

garrett16r opened the pull request:
This turned out to be far easier than I had expected. Discord.js-commando handles multi-word arguments internally, so there was no need to worry about converting the argument for a region like "US East" to an array, then back to a string, etc.

In constants.js, I had to make the color roles and regional roles use separate consts because of the way that the code confirmed that the user only had one color role before. While it does create one extra check in a couple if statements in giverole.js, I think the benefit of only needing to modify a few lines of code rather than creating a whole system of checks to see if the user only has one color role _and_ only one regional role due to using one `AVAILABLE_ROLES` const is worth it.

Finally, I switched `r!roles` from using an ugly plaintext output to using an embed with much nicer formatting.

Commit:
Begin work on regional roles

- Fixed formatting for r!roles command output
- Added regional roles to r!roles
- Began adding regional roles to constants.js

Commit:
Make r!roles use an embed

+ Rather than using a plain text message with bad formatting, r!roles now sends an embed with a nice table-like layout

Commit:
Update Readme.md

Commit:
Added regional roles

+ Added regional roles to constants.js as a separate constant from color roles.
+ Adjusted logic to check for regional roles as well as color roles.
+ Added a general Europe role for people who are not in one of the specific locations of the other European roles.

Commit:
Delete start.bat

Commit:
Merge pull request #1 from Syntthetix/syn-region-roles-dev

Syn region roles dev

Commit:
Optimized roles.js

- Code now checks that the user is in the correct channel before anything else in order to prevent running unnecessary code.

Commit:
Split logical checks into multiple if statements in giverole.js

+ Rather than checking four different criteria in a single if statement, the code now has two separate if statements to handle smaller checks. This change is largely stylistic and for readability.

Commit:
Merge pull request #2 from Syntthetix/master

Update dev branch to contain latest changes

Commit:
Cleaned up giverole.js

- Moved checks for whether or not a user can claim a role into the claimRole() method in util.js.
- Found that the check that would tell the user if a role they tried to claim doesn't exist could not actually work. This was also an unnecessary piece of code, so it was removed.

Commit:
Merge pull request #3 from Syntthetix/syn-region-roles-dev

Syn region roles dev

garrett16r commented:
(I know you're seeing weird merges in there, it's just because of the fact that I made the PR from my master branch by accident, but I've been pushing things to my dev branch. One day I'll get used to everything with github.)

causztic reviewed (commented):
A couple of things!
- If a function is only used in one file, usually there is no need to move it to a "general" file like `util.js`, it is fine to leave it in the same file.
- Moving one lump of code of high complexity into a function is usually not good refactoring (it's like moving a large pile of laundry from the washing machine to your own room.. the pile is still there). What you want to do here is to break down the complexity into smaller, readable functions! For example, the claimRole function can be split into `giveMemberRole`, `giveColorRole` and `giveRegionRole` responsibilities. (so the laundry gets split up into different rooms so that you can find each piece of clothing better based on who it belongs to)

Commit:
Restructured giverole.js

- Moved claimRole() out of util.js and broke it into giveMemberRole(), giveColorRole(), giveRegionRole(), and grantRole() inside giverole.js.
- Fixed rmrole.js, as the check to see if the role provided by the user exists was using the wrong argument, resulting in a discord.js error message if the role didn't exist.

causztic commented on lib/commands/misc/giverole.js:
```diff
@@ -24,46 +24,80 @@ module.exports = class GiveRoleCommand extends Command {
+			}
+
+		} else {
+		  msg.reply("You have already claimed that role! Use r!rmrole [role name] and try again.");
```
Your if condition checks that "the user doesn't already have a color role" and "is a Member", but the message only caters for the first case! Let's change the if statement. You can also look into checking for "is a Member" before calling this method.

causztic commented on lib/commands/misc/giverole.js:
```diff
@@ -24,46 +24,80 @@ module.exports = class GiveRoleCommand extends Command {
+
+	giveRegionRole(msg, role, channel, userRoles, memberRole) {
+		// Confirms that the user doesn't already have the region role and is a Member
+		if (!userRoles.has(role.id) && userRoles.has(memberRole.id)) {
```
same as above!

causztic reviewed (changes requested):
some cleaning up and making some messages more accurate!

causztic commented on lib/commands/misc/giverole.js:
```diff
@@ -24,46 +24,80 @@ module.exports = class GiveRoleCommand extends Command {
-        const channel = this.client.channels.get(getChannel(GENERAL_ID));
-        channel.send(`Welcome to the server, ${msg.author}! <:wutGiga:297897855727697921>`);
-      }
+    } else if (COLOR_ROLES.includes(roleArg) || REGION_ROLES.includes(roleArg)) {
```
The if statements here can be cleaner, maybe something like:

first conditional block:
1. if the argument is member, call `giveMemberRole` and return
2. otherwise, if the user is a member, check that it is in a command channel
3. if not, return the command
4. if the user is in a command channel, check if the user is a member.
5. if the user is not a member, return a message

second conditional block:
1. If it is a COLOR_ROLE, `giveColorRole`
2. else if it is a REGION_ROLE, `giveRegionRole`
3. otherwise, it is not a valid role.

Commit:
Clean up giverole.js

- Based on PR comments, cleaned up the logical flow of giverole.js to avoid reused code and improve readability.

causztic commented on lib/commands/misc/giverole.js:
```diff
@@ -24,46 +24,84 @@ module.exports = class GiveRoleCommand extends Command {
+
+	giveColorRole(msg, role, channel, userRoles) {
+		// Confirms that the user doesn't already have the color role
+		if (!userRoles.has(role.id)) {
```
one final change! `giveColorRole` and `giveRegionRole` both checks that the user has the role or not, you could move this out to the `run` function, before line 99 where you `if (COLOR_ROLES.includes(roleArg)) {`

this way you only check it once instead of two times in total for each of those functions.

causztic reviewed (changes requested):
Looks good now! just *one* final change :)

Commit:
Optimize of giverole.js

- Moved check for whether the user already has a role they are trying to claim out of the giveColorRole() and giveRegionRole() methods to prevent reusing code.

Commit:
Update giverole.js

Commit:
Update giverole.js

+ Added check to confirm that a role actually exists before running any other checks.

Commit:
Changed names of two roles

- "Scandinavian" role was renamed to "Scandinavia"
- "Southeast Asia" role was renamed to "SEA"

causztic reviewed (approved):
![QuerulousFalseBream-size_restricted](https://user-images.githubusercontent.com/2081441/76000799-2e255b80-5f3f-11ea-8da7-4ee5ac2a64b0.gif)
